### ISSUE TITLE
perf: Amdahl v3 — training step optimization roadmap for real models

### ISSUE THREAD

**Issue by teilomillet** [teilomillet/textpolicy#41]
## Context

Follow-up to #23 (Amdahl v2, now closed). All v2 optimizations are verified and working:
- Batched logprobs (#27): `loss_and_grad` dropped from 66.5% → 23.6%
- Batched generation (#26): rollout throughput improved
- Buffer growth (#35): training time stays constant
- KV-cache sharing (#29): measured, not worth implementing (< 1% savings)

### Current profile (Qwen3-0.6B, 4 episodes × 128 tokens)

```
FULL STEP: 2,958 ms
├─ Rollout (batch_generate_tokens):  2,354 ms (79.6%)  ← KV-cache active, 5.4x faster than without
└─ Training:                           604 ms (20.4%)  ← true cost, unprofiled
   ├─ loss_and_grad:                   148 ms (24.5%)
   ├─ grad_clip:                       412 ms (68.2%)  ← NEW DOMINANT BOTTLENECK
   └─ optimizer_update:                 44 ms ( 7.3%)
```

**Key correction**: Earlier synthetic benchmarks (TinyLM, 50K params) showed grad_clip as a profiling artifact. With Qwen3-0.6B (596M params), grad_clip is a **real 62.8% overhead** — two full traversals over the parameter tree.

### Phase 1: Training step optimization (confirmed bottlenecks)

- [ ] #39 perf: optimize gradient clipping for large models (62.8% overhead → target ≤20%)
- [ ] #40 perf: enable mx.compile by default (16% speedup on real models)

### Phase 2: Investigate bottleneck under different training modes

- [ ] #42 perf: profile LoRA training — does frozen-base eliminate the grad_clip bottleneck?
- [ ] #43 perf: benchmark generation with quantized models (4-bit inference)

### Phase 3: Measurement infrastructure

- [ ] #44 test: add real-model regression benchmark (Qwen3-0.6B litmus tests)

## Estimated impact

```
Current:     Training = 1,069 ms (profiled) / 604 ms (unprofiled)
After #39:   Training ~ 450 ms (skip unnecessary scaling, ~30% faster)
After #40:   Training ~ 380 ms (compile fusion on top, ~16% more)
Combined:    Training ~ 350-400 ms (target: ~1.7x training speedup)

Full step:   2,958 ms → ~2,700 ms (8-9% end-to-end improvement on full model)
```

With LoRA (#42) and quantization (#43), the profile may shift dramatically:
- LoRA: grad_clip may drop from 412ms to <10ms (only ~1-5M trainable params)
- 4-bit quantization: rollout may drop from 2,354ms to ~800ms (4x less memory bandwidth)
- **Combined LoRA + 4-bit: full step could go from ~3s to ~1s (3x overall speedup)**

## Execution rules (inherited from #23)

- Every perf PR must include both correctness parity tests AND perf delta evidence
- Profile before optimizing — don't optimize blind
- Preserve existing invariants: flat 1D logprobs, episode_lengths consistency
- No regression in 539+ test suite
- **Use real models for benchmarks** — TinyLM results were misleading for grad_clip

**Comment by teilomillet** [comment]
## Resolved — all sub-tasks complete, LoRA profiling shows fundamentally different bottleneck landscape

### Sub-task status

| # | Issue | Status |
|---|-------|--------|
| #39 | grad_clip 62.8% overhead | ✅ Closed — lazy clipping implemented, 0.1% under LoRA |
| #40 | mx.compile for training | ✅ Closed — auto-compile with 1M param threshold |
| #42 | Profile LoRA training | ✅ Closed — profiling confirms grad_clip is negligible under LoRA |
| #43 | Benchmark quantized models | ✅ Closed — 4-bit inference benchmarked |
| #44 | Real-model regression tests | ✅ Closed — litmus tests in test_real_model_perf.py |

### End-to-end Amdahl breakdown (Trinity-Nano-Preview, LoRA, 5 steps)

```
rollout_collect     78.89s  ( 83.1%)  Amdahl limit if perfect: 5.90x
train               16.08s  ( 16.9%)  Amdahl limit if perfect: 1.20x

Trainer-internal:
  loss_and_grad       15.66s  ( 97.4%)
  optimizer_update     0.35s  (  2.2%)
  grad_clip            0.02s  (  0.1%)
```

### Key insight
The entire training step optimization roadmap was profiled on **Qwen3-0.6B with full-model training** (596M trainable params). Our production setup uses **LoRA on Trinity-Nano-Preview** where only 0.07% of params are trainable. The bottleneck has shifted entirely to rollout collection (83.1%) — further training-step optimizations have diminishing returns (Amdahl limit: 1.20x).

All 612 tests pass, 0 regressions.
